include/bm/bm_sim/nic.h

```diff
@@ -0,0 +1,1104 @@
+//! bm::Nic class and its dataplane APIs. However, because we offer unified
+//! runtime APIs, you will have to use a context id of `0` when programming the
+//! tables, even when your nic class inherits from bm::Nic and not
+//! bm::NicWContexts.
+//! The simple switch target only supports one bm::Context and inherits from
+//! bm::Nic.
+//!
+//! When subclassing on of these two classes, you need to remember to implement
+//! the two pure virtual functions:
+//! bm::NicWContexts::receive_(int port_num, const char *buffer, int len) and
+//! bm::NicWContexts::start_and_return_(). Your receive() implementation will
+//! be called for you every time a new packet is received by the device. In your
+//! start_and_return() function, you are supposed to start the different
+//! processing threads of your target switch and return immediately. Note that
+//! start_and_return() should not be mandatory per se (the target designer could
+//! do the initialization any way he wants, even potentially in the
+//! constructor). However, we have decided to keep it around for now.
+//!
+//! Both nic classes support live swapping of P4-JSON configurations. To
+//! enable it you need to provide the correct flag to the constructor (see
+//! bm::NicWContexts::NicWContexts()). Swaps are ordered through the
+//! runtime interfaces. We ensure that during the actual swap operation
+//! (bm::NicWContexts::do_swap() method), there is no Packet instance
+//! inflight, which we achieve using the process_packet_mutex mutex). The final
+//! step of the swap is to call bm::NicWContexts::swap_notify_(), which
+//! targets can override if they need to perform some operations as part of the
+//! swap. Targets are guaranteed that no Packet instances exist as that
+//! time. Note that swapping configurations may invalidate pointers that you are
+//! still using, and it is your responsibility to refresh them.
+
+#ifndef BM_BM_SIM_NIC_H_
+#define BM_BM_SIM_NIC_H_
+
+#include <condition_variable>
+#include <iosfwd>
+#include <memory>
+#include <set>
+#include <string>
+#include <typeindex>
+#include <typeinfo>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include <boost/thread/shared_mutex.hpp>
+
+#include "action_profile.h"
+#include "context.h"
+#include "dev_mgr.h"
+#include "device_id.h"
+#include "learning.h"
+#include "lookup_structures.h"
+#include "phv_source.h"
+#include "queue.h"
+#include "runtime_interface.h"
+#include "target_parser.h"
+
+namespace bm {
+
+class OptionsParser;
+class Packet;
+
+// multiple inheritance in accordance with Google C++ guidelines:
+// "Multiple inheritance is allowed only when all superclasses, with the
+// possible exception of the first one, are pure interfaces. In order to ensure
+// that they remain pure interfaces, they must end with the Interface suffix."
+//
+//! Base class for a switch implemenattion where multi-context support is
+//! required.
+class NicWContexts : public DevMgr, public RuntimeInterface {
```

Comment:
we should really avoid introducing this object as we are just duplicating a lot of code. If some changes are needed to `SwitchWContexts` to accommodate the PNA use case, we should first consider making these changes, instead of introducing `NicWContexts`.

Can you provide some justification for the introduction of this class?

Reply:
Yes, code wise, both the _NicWContexts_ and _SwitchWContexts_ are the same.

Within the _targets/pna_nic_ directory, the _PnaNic_ class is required to inherit from the _Switch_ class. However, considering that a Nic and a Switch are different physical entities. I just felt odd for the _PnaNic_ to inherit directly from the _Switch_. So, I created a duplicate _Nic_ class. But yes, _Nic_ is a duplicate of _Switch_ class and _NicWContexts_ is a duplicate of _SwitchWContexts_ class. 

If the community agrees to eliminate these duplicates and allowing _PnaNic_ class to inherit from the Switch class, we will change accordingly.

Please provide feedback.

Reply:
A NIC following the PNA architecture is not that different from a switch. The "switch" terminology is not totally out of place here. Duplicating the code would be the worst possible solution here, it creates maintenance issues.

Reply:
Maybe move all/most of the code to a common base `Device` class that `Nic` and `Switch` both inherit from?

Reply:
Here are two possible approaches, and what I see as their tradeoffs:

(a) Keep the existing class name `Switch`, and make `Nic` a subclass of it.

Advantages:
+ avoids the code duplication that Antonin commented on.

Disadvantages:
+ It seems like it needs a comment noting that yes, we agree that a class hierarchy where all instances of `Nic` are also instances of `Switch` might not be ideal, it was a reasonable choice given the alternatives.  One might not even think of it as a poor choice of hierarchy if you think of things like Open Vswitch being implemented in NICs: Why is it called V_switch_?  Well, probably because it is switching packets from a set of interfaces, and can end up sending them to 0, 1, or more other interfaces, i.e. it is switching packets between interfaces.

(b) Introduce a new class `Device` that is just a rename of the current `Switch` class, and make a new `Switch` class that is a subclass of it, and a new `Nic` class that is also a subclass of `Device`.

Advantages:
+ completely avoids any mental confusion anyone might have about whether a NIC is a kind of Switch, or not.

Disadvantages:
+ _If_ this requires a lot of little changes throughout the code, one might consider that a disadvantage.  I have not evaluated for certain whether it would require that.

I do not have a strong preference for either (a) or (b), myself, but agree with Antonin's review that approach (c) copy and paste class `Switch` to make a new class `Nic` does seem like significant code duplication.

Reply:
Both options seems reasonable.

option (a): making `Nic` subclass of `Switch`; is an easy approach. But, if we are going to follow option (a), It will be better to follow @antoninbas suggestion and make `PnaNic` class directly inherit `Switch` class.

Following option (b) as @ChrisDodd suggested: Introducing new class `Device`; would need to make few changes in the below mentioned files:
| Source | Number of times `Switch` or `SwitchWContexts` classes are mentioned |
|--------|--------|
| targets/simple_router/simple_router.cpp | 2 |
| targets/l2_switch/l2_switch.cpp | 1 |
| targets/simple_switch_grpc/switch_runner.cpp | 1 |
| targets/psa_switch/psa_switch.cpp | 1 |
| targets/simple_switch/runner.cpp | 1 | 
| include/bm/bm_sim/packet.h | 2 |
| include/bm/bm_sim/context.h | only comments |
| include/bm/bm_sim/lookup_structures.h | 2 |
| include/bm/bm_runtime/bm_runtime.h | 1 |
| include/Makefile.am | 1 |
| PI/src/common.h | 2 |
| PI/src/pi_mc_imp.cpp | 1 |
| PI/src/pi_meter_imp.cpp | 1 |
| PI/src/pi.cpp | 1 |
| src/bm_sim/Makefile.am | 1 |
| src/bm_runtime/SimplePreLAG_server.cpp | 3 |
| src/bm_runtime/server.cpp | 5 |
| src/bm_runtime/Standard_server.cpp | 4 |
| src/bm_runtime/SimplePre_server.cpp | 3 |
| include/bm/bm_sim/switch.h | class changes |
| src/bm_sim/switch.cpp | class changes |
| Need to change documentation files (.md files) | |
| Few more changes in tests | |

But, As @jafingerhut mentioned, following option (b) will avoid any mental confusion between `Switch` and `Nic` in the future.

If the community thinks the above number changes are reasonable and not too many, I will prefer going with option (b). Otherwise, @antoninbas suggestion is a better approach.

Reply:
Or Just name changing _Switch_ class to _Device_ class would be fine. Right?

Reply:
I suspect there will be many more significant decisions than this to make in order to enable Nic to work in BMv2 than this decision, and it could easily be 3-4 days between rounds of this conversation.

What do you think of the option of going with Antonin Bas's suggestion for now, and maybe we can revisit later whether changing class name `Switch` to `Device` is preferable?

Reply:
Agree. I'd prefer that we aim to get a reasonable implementation in ASAP, and iterate on improving it later.

Reply:
Ok. 
I will start by splitting the PR. The first part will include a basic implementation of `pna_nic`. This initial version will serve as a skeleton and will support basic P4 PNA programs (without tables).